File: telegram-bot/src/favourites.js
import { db } from "./db.js";

const listStmt = db.prepare(
  "SELECT stop_code AS code, stop_name AS name FROM favourites WHERE chat_id = ? ORDER BY stop_name"
);
const isFavStmt = db.prepare(
  "SELECT 1 FROM favourites WHERE chat_id = ? AND stop_code = ?"
);
const addStmt = db.prepare(
  `INSERT INTO favourites (chat_id, stop_code, stop_name, created_at)
   VALUES (?, ?, ?, ?)
   ON CONFLICT(chat_id, stop_code) DO UPDATE SET stop_name = excluded.stop_name`
);
const removeStmt = db.prepare(
  "DELETE FROM favourites WHERE chat_id = ? AND stop_code = ?"
);

export function listFavourites(chatId) {
  return listStmt.all(chatId);
}

export function isFavourite(chatId, code) {
  return !!isFavStmt.get(chatId, code);
}

export function addFavourite(chatId, code, name) {
  addStmt.run(chatId, code, name, Date.now());
}

export function removeFavourite(chatId, code) {
  removeStmt.run(chatId, code);
}

export function toggleFavourite(chatId, code, name) {
  if (isFavourite(chatId, code)) {
    removeFavourite(chatId, code);
    return false;
  }
  addFavourite(chatId, code, name);
  return true;
}

File: telegram-bot/src/favourites.py
import time

from .db import db


def list_favourites(chat_id: int) -> list:
    return db.execute(
        "SELECT stop_code AS code, stop_name AS name FROM favourites WHERE chat_id = ? ORDER BY stop_name",
        (chat_id,),
    ).fetchall()


def is_favourite(chat_id: int, code: str) -> bool:
    return db.execute(
        "SELECT 1 FROM favourites WHERE chat_id = ? AND stop_code = ?", (chat_id, code)
    ).fetchone() is not None


def add_favourite(chat_id: int, code: str, name: str) -> None:
    with db:
        db.execute(
            """
            INSERT INTO favourites (chat_id, stop_code, stop_name, created_at)
            VALUES (?, ?, ?, ?)
            ON CONFLICT(chat_id, stop_code) DO UPDATE SET stop_name = excluded.stop_name
            """,
            (chat_id, code, name, int(time.time() * 1000)),
        )


def remove_favourite(chat_id: int, code: str) -> None:
    with db:
        db.execute("DELETE FROM favourites WHERE chat_id = ? AND stop_code = ?", (chat_id, code))


def toggle_favourite(chat_id: int, code: str, name: str) -> bool:
    if is_favourite(chat_id, code):
        remove_favourite(chat_id, code)
        return False
    add_favourite(chat_id, code, name)
    return True
